iopipe/agent.py

```diff
@@ -0,0 +1,231 @@
+                timeout_duration = max([0, timeout_duration])
+
+                # Maximum execution time is 10 minutes, make sure timeout doesn't exceed that minus the timeout window
+                timeout_duration = min([timeout_duration, 60 * 60 * 10 - self.config['timeout_window']])
+
+                logger.debug('Setting timeout duration to %s' % timeout_duration)
+
+                # Using signal.setitimer instead of signal.alarm because the latter only accepts integers and we want to
+                # be able to timeout at millisecond granularity
+                signal.setitimer(signal.ITIMER_REAL, timeout_duration)
+
+            result = None
+
+            try:
+                result = func(event, context)
+            except Exception as e:
+                self.run_hooks('post:invoke', event=event, context=context)
+
+                # This prevents this block from being executed a second time in the event that a timeout occurs and an
+                # exception is subsequently raised within the handler
+                if self.report.sent is False:
+                    self.report.prepare(e)
+                    self.run_hooks('pre:report')
+                    self.report.send()
+                    self.run_hooks('post:report')
+
+                raise e
+            else:
+                self.run_hooks('post:invoke', event=event, context=context)
+                self.report.prepare()
+                self.run_hooks('pre:report')
+                self.report.send()
+                self.run_hooks('post:report')
+            finally:
+                signal.setitimer(signal.ITIMER_REAL, 0)
+                self.wait_for_futures()
+
+            return result
+
+        return wrapped
+
+    decorator = __call__
+
+    def handle_timeout(self, signum, frame):
+        """
+        Catches a timeout (SIGALRM) and sends the report before actual timeout occurs.
+
+        The signum and frame parameters are passed by the signal module to this handler.
+
+        :param signum: The signal number being handled.
+        :param frame: The stack frame when signal was raised.
+        """
+        logger.debug('Function is about to timeout, sending report')
+        self.report.prepare(TimeoutError('Timeout Exceeded.'), frame)
+        self.run_hooks('pre:report')
+        self.report.send()
+        self.run_hooks('post:report')
+        self.wait_for_futures()
+
+    def load_plugins(self, plugins):
+        """
+        Loads plugins that match the `Plugin` interface and are instantiated.
+
+        :param plugins: A list of plugin instances.
+        """
+
+        def instantiate(plugin):
+            return plugin() if inspect.isclass(plugin) else plugin
+
+        loaded_plugins = []
+        plugins_seen = []
+
+        for plugin in reversed(plugins):
+            if not is_plugin(plugin) or plugin.name in plugins_seen:
+                continue
+            # Build the plugins list in reverse to restore original order
```

Comment:
It looks like this comment belongs above 186, and it'd be helpful to clarify that you're using the "reverse order, so that user plugins override kitchen sink* plugins" *use a name that makes sense

Reply:
No, the comment is in the correct location, as that's where the list is being built. Added an additional comment above the loop to explain why we're iterating over the plugins in reverse order.

Reply:
Sounds good